File: engine/src/core/persistence/SaveGameManager.cpp
#include "core/persistence/SaveGameManager.h"

#include <filesystem>
#include <fstream>
#include <sstream>

namespace subspace {

static const std::string kSaveExtension = ".save";

SaveGameManager& SaveGameManager::Instance()
{
    static SaveGameManager instance;
    return instance;
}

bool SaveGameManager::SaveGame(const SaveGameData& data, const std::string& fileName)
{
    namespace fs = std::filesystem;
    fs::create_directories(_saveDirectory);

    std::string path = _saveDirectory + "/" + fileName + kSaveExtension;
    std::ofstream file(path);
    if (!file.is_open()) return false;

    // Header
    file << "[HEADER]\n";
    file << "saveName=" << data.saveName << "\n";
    file << "saveTime=" << data.saveTime << "\n";
    file << "version=" << data.version << "\n";
    file << "galaxySeed=" << data.galaxySeed << "\n";

    // Game state
    file << "[GAMESTATE]\n";
    for (const auto& [key, value] : data.gameState) {
        file << key << "=" << value << "\n";
    }

    // Entities
    for (const auto& entity : data.entities) {
        file << "[ENTITY]\n";
        file << "id=" << entity.entityId << "\n";
        file << "name=" << entity.entityName << "\n";
        file << "active=" << (entity.isActive ? "true" : "false") << "\n";

        for (const auto& comp : entity.components) {
            file << "[COMPONENT]\n";
            file << "type=" << comp.componentType << "\n";
            for (const auto& [key, value] : comp.data) {
                file << key << "=" << value << "\n";
            }
        }
    }

    return file.good();
}

bool SaveGameManager::LoadGame(const std::string& fileName, SaveGameData& outData)
{
    std::string path = _saveDirectory + "/" + fileName + kSaveExtension;
    std::ifstream file(path);
    if (!file.is_open()) return false;

    outData = SaveGameData{};

    enum class Section { None, Header, GameState, Entity, Component };
    Section section = Section::None;

    std::string line;
    while (std::getline(file, line)) {
        if (line.empty()) continue;

        // Detect section markers
        if (line == "[HEADER]") { section = Section::Header; continue; }
        if (line == "[GAMESTATE]") { section = Section::GameState; continue; }
        if (line == "[ENTITY]") {
            section = Section::Entity;
            outData.entities.emplace_back();
            continue;
        }
        if (line == "[COMPONENT]") {
            section = Section::Component;
            if (!outData.entities.empty()) {
                outData.entities.back().components.emplace_back();
            }
            continue;
        }

        // Parse key=value
        auto pos = line.find('=');
        if (pos == std::string::npos) continue;
        std::string key = line.substr(0, pos);
        std::string value = line.substr(pos + 1);

        switch (section) {
        case Section::Header:
            if (key == "saveName") outData.saveName = value;
            else if (key == "saveTime") outData.saveTime = value;
            else if (key == "version") outData.version = value;
            else if (key == "galaxySeed") {
                try { outData.galaxySeed = std::stoi(value); }
                catch (...) { outData.galaxySeed = 0; }
            }
            break;
        case Section::GameState:
            outData.gameState[key] = value;
            break;
        case Section::Entity:
            if (outData.entities.empty()) break;
            if (key == "id") {
                try { outData.entities.back().entityId = std::stoull(value); }
                catch (...) { outData.entities.back().entityId = 0; }
            }
            else if (key == "name") outData.entities.back().entityName = value;
            else if (key == "active") outData.entities.back().isActive = (value == "true");
            break;
        case Section::Component:
            if (outData.entities.empty() || outData.entities.back().components.empty()) break;
            if (key == "type") outData.entities.back().components.back().componentType = value;
            else outData.entities.back().components.back().data[key] = value;
            break;
        default:
            break;
        }
    }

    return true;
}

std::vector<SaveGameInfo> SaveGameManager::ListSaveGames() const
{
    namespace fs = std::filesystem;
    std::vector<SaveGameInfo> result;

    if (!fs::exists(_saveDirectory)) return result;

    for (const auto& entry : fs::directory_iterator(_saveDirectory)) {
        if (!entry.is_regular_file()) continue;
        if (entry.path().extension() != kSaveExtension) continue;

        // Read just the header to populate info
        std::ifstream file(entry.path());
        if (!file.is_open()) continue;

        SaveGameInfo info;
        info.fileName = entry.path().stem().string();

        std::string line;
        bool inHeader = false;
        while (std::getline(file, line)) {
            if (line == "[HEADER]") { inHeader = true; continue; }
            if (!line.empty() && line[0] == '[') break;
            if (!inHeader) continue;

            auto pos = line.find('=');
            if (pos == std::string::npos) continue;
            std::string key = line.substr(0, pos);
            std::string value = line.substr(pos + 1);

            if (key == "saveName") info.saveName = value;
            else if (key == "saveTime") info.saveTime = value;
            else if (key == "version") info.version = value;
        }

        result.push_back(std::move(info));
    }

    return result;
}

bool SaveGameManager::DeleteSave(const std::string& fileName)
{
    namespace fs = std::filesystem;
    std::string path = _saveDirectory + "/" + fileName + kSaveExtension;
    return fs::remove(path);
}

bool SaveGameManager::QuickSave(const SaveGameData& data)
{
    return SaveGame(data, "quicksave");
}

std::string SaveGameManager::GetSaveDirectory() const
{
    return _saveDirectory;
}

void SaveGameManager::SetSaveDirectory(const std::string& dir)
{
    _saveDirectory = dir;
}

} // namespace subspace

File: AvorionLike/Core/Persistence/SaveGameManager.cs
using System.Text.Json;
using AvorionLike.Core.Logging;

namespace AvorionLike.Core.Persistence;

/// <summary>
/// Save game data structure
/// </summary>
public class SaveGameData
{
    public string SaveName { get; set; } = "";
    public DateTime SaveTime { get; set; }
    public string Version { get; set; } = "1.0.0";
    public int GalaxySeed { get; set; }
    public Dictionary<string, object> GameState { get; set; } = new();
    public List<EntityData> Entities { get; set; } = new();
}

/// <summary>
/// Serialized entity data
/// </summary>
public class EntityData
{
    public Guid EntityId { get; set; }
    public string EntityName { get; set; } = "";
    public bool IsActive { get; set; }
    public List<ComponentData> Components { get; set; } = new();
}

/// <summary>
/// Serialized component data
/// </summary>
public class ComponentData
{
    public string ComponentType { get; set; } = "";
    public Dictionary<string, object> Data { get; set; } = new();
}

/// <summary>
/// Manages saving and loading game state
/// </summary>
public class SaveGameManager
{
    private static SaveGameManager? _instance;
    private readonly string _saveDirectory;

    public static SaveGameManager Instance
    {
        get
        {
            _instance ??= new SaveGameManager();
            return _instance;
        }
    }

    private SaveGameManager()
    {
        _saveDirectory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
            "Codename-Subspace",
            "Saves"
        );

        // Ensure save directory exists
        if (!Directory.Exists(_saveDirectory))
        {
            Directory.CreateDirectory(_saveDirectory);
            Logger.Instance.Info("SaveGameManager", $"Created save directory: {_saveDirectory}");
        }
    }

    /// <summary>
    /// Get the saves directory path
    /// </summary>
    public string GetSaveDirectory()
    {
        return _saveDirectory;
    }

    /// <summary>
    /// List all available save files
    /// </summary>
    public List<SaveGameInfo> ListSaveGames()
    {
        var saves = new List<SaveGameInfo>();

        try
        {
            var files = Directory.GetFiles(_saveDirectory, "*.save");
            
            foreach (var file in files)
            {
                try
                {
                    var json = File.ReadAllText(file);
                    var saveData = JsonSerializer.Deserialize<SaveGameData>(json);
                    
                    if (saveData != null)
                    {
                        saves.Add(new SaveGameInfo
                        {
                            FileName = Path.GetFileName(file),
                            SaveName = saveData.SaveName,
                            SaveTime = saveData.SaveTime,
                            Version = saveData.Version
                        });
                    }
                }
                catch (Exception ex)
                {
                    Logger.Instance.Warning("SaveGameManager", $"Failed to read save file {file}: {ex.Message}");
                }
            }
        }
        catch (Exception ex)
        {
            Logger.Instance.Error("SaveGameManager", "Failed to list save games", ex);
        }

        return saves.OrderByDescending(s => s.SaveTime).ToList();
    }

    /// <summary>
    /// Save game state to file
    /// </summary>
    public bool SaveGame(SaveGameData saveData, string fileName)
    {
        try
        {
            if (!fileName.EndsWith(".save"))
            {
                fileName += ".save";
            }

            var filePath = Path.Combine(_saveDirectory, fileName);
            
            saveData.SaveTime = DateTime.UtcNow;
            
            var options = new JsonSerializerOptions
            {
                WriteIndented = true
            };
            
            var json = JsonSerializer.Serialize(saveData, options);
            File.WriteAllText(filePath, json);
            
            Logger.Instance.Info("SaveGameManager", $"Game saved to: {filePath}");
            return true;
        }
        catch (Exception ex)
        {
            Logger.Instance.Error("SaveGameManager", $"Failed to save game to {fileName}", ex);
            return false;
        }
    }

    /// <summary>
    /// Load game state from file
    /// </summary>
    public SaveGameData? LoadGame(string fileName)
    {
        try
        {
            if (!fileName.EndsWith(".save"))
            {
                fileName += ".save";
            }

            var filePath = Path.Combine(_saveDirectory, fileName);
            
            if (!File.Exists(filePath))
            {
                Logger.Instance.Warning("SaveGameManager", $"Save file not found: {filePath}");
                return null;
            }

            var json = File.ReadAllText(filePath);
            var saveData = JsonSerializer.Deserialize<SaveGameData>(json);
            
            if (saveData == null)
            {
                Logger.Instance.Error("SaveGameManager", $"Failed to deserialize save file: {filePath}");
                return null;
            }

            Logger.Instance.Info("SaveGameManager", $"Game loaded from: {filePath}");
            return saveData;
        }
        catch (Exception ex)
        {
            Logger.Instance.Error("SaveGameManager", $"Failed to load game from {fileName}", ex);
            return null;
        }
    }

    /// <summary>
    /// Delete a save file
    /// </summary>
    public bool DeleteSave(string fileName)
    {
        try
        {
            if (!fileName.EndsWith(".save"))
            {
                fileName += ".save";
            }

            var filePath = Path.Combine(_saveDirectory, fileName);
            
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
                Logger.Instance.Info("SaveGameManager", $"Deleted save file: {filePath}");
                return true;
            }
            
            return false;
        }
        catch (Exception ex)
        {
            Logger.Instance.Error("SaveGameManager", $"Failed to delete save {fileName}", ex);
            return false;
        }
    }

    /// <summary>
    /// Create a quick save with auto-generated name
    /// </summary>
    public bool QuickSave(SaveGameData saveData)
    {
        var fileName = $"quicksave_{DateTime.Now:yyyyMMdd_HHmmss}.save";
        saveData.SaveName = "Quick Save";
        return SaveGame(saveData, fileName);
    }

    /// <summary>
    /// Get the most recent save file
    /// </summary>
    public SaveGameInfo? GetMostRecentSave()
    {
        var saves = ListSaveGames();
        return saves.FirstOrDefault();
    }
}

/// <summary>
/// Information about a save game file
/// </summary>
public class SaveGameInfo
{
    public string FileName { get; set; } = "";
    public string SaveName { get; set; } = "";
    public DateTime SaveTime { get; set; }
    public string Version { get; set; } = "";
}
